RavenDB-15776 - Add timings for Lucene Merge step

grisha-kotler opened the pull request:
https://issues.hibernatingrhinos.com/issue/RavenDB-15776

grisha-kotler committed:
RavenDB-15776 - Add timings for Lucene Merge step

grisha-kotler committed:
RavenDB-15776 - add copy file stream to the commit stats

ayende commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/CustomSerialMergeScheduler.cs:
```diff
@@ -0,0 +1,26 @@
+        {
+            using (_commitStats?.For(IndexingOperation.Lucene.Merge))
+            {
+                base.Merge(writer, state);
```
Can we provide additional information here? In particular, what was merged, context, etc?
https://github.com/ravendb/lucenenet/blob/3.0.3-netcore/src/Lucene.Net/Index/MergePolicy.cs#L72

ayende replied:
Number of items that are being merged is also interesting
Also, I think we should be able to set a time limit, no?
So we'll only allow a certain amount of time to go here, then force a merge if we need to when the index is idle?

ppekrol replied:
We have IndexingRunStats so it should not be a problem. We can just copy implementation of SerialMergeScheduler (https://github.com/ravendb/lucenenet/blob/3.0.3-netcore/src/Lucene.Net/Index/SerialMergeScheduler.cs) and write our own that will accumulate this.

What would be a reasonable time limit? What is the impact of not doing merges?

grisha-kotler replied:
Setting a time limit doesn't improve performance. It just delays the merge to the next batch.
I added the merge details.

ppekrol commented on src/Raven.Client/Documents/Indexes/IndexingOperation.cs:
```diff
@@ -30,12 +30,14 @@ internal static class Lucene
         internal static class Storage
         {
             public const string Commit = "Storage/Commit";
+            public const string CopyFileStream = "Voron/CopyFileStream";
```
`Storage/CopyFileStream` ?

grisha-kotler replied:
removed since it doesn't consume much time

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexWriter.cs:
```diff
@@ -62,10 +64,12 @@ public void DeleteDocuments(Term term, IState state)
             try
             {
+                _directory.SetCommitStats(commitStats);
+                _mergeScheduler.SetCommitStats(commitStats);
```
I think it would be better to release this after the commit, in some using so if something goes wrong we will not stay with 'leftovers'

grisha-kotler replied:
done

grisha-kotler committed:
RavenDB-15776 - don't wait for merges in dispose (we already executed the merge in the commit stage)

grisha-kotler committed:
RavenDB-15776 - track ApplyDeletes stats
- track LuceneMerge stats

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/CustomIndexWriter.cs:
```diff
@@ -0,0 +1,45 @@
+
+namespace Raven.Server.Documents.Indexes.Persistence.Lucene
+{
+    public class CustomIndexWriter : IndexWriter
```
Can we have a better name here? Maybe `TimeTrackingIndexWriter` ? Something that would indicate that that it updates the timings of the stats

grisha-kotler replied:
done

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/CustomSerialMergeScheduler.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace Raven.Server.Documents.Indexes.Persistence.Lucene
+{
+    public class CustomSerialMergeScheduler : MergeScheduler
```
Same here - `TimeTrackingSerialMergeScheduler` ?

grisha-kotler replied:
done

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexWriter.cs:
```diff
@@ -160,14 +166,16 @@ private void CreateIndexWriter(IState state)
-            _indexWriter.SetMergeScheduler(new SerialMergeScheduler(), state);
+
+            _mergeScheduler = new CustomSerialMergeScheduler();
+            _indexWriter.SetMergeScheduler(_mergeScheduler, state);
```
Can we require to pass `_mergeScheduler ` to the constructor of `new CustomIndexWriter` instead of setting it this way?

grisha-kotler replied:
done

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexWriter.cs:
```diff
@@ -187,7 +195,7 @@ private void DisposeIndexWriter(bool waitForMerges = true)
             try
             {
-                writer.Dispose(waitForMerges);
+                writer.Dispose(waitForMerges: false);
```
Did you change the default behavior?

grisha-kotler replied:
waitingForMerges: true triggers a new index merge (if needed)
We are already doing the merge during the commit stage.

ppekrol replied:
It could be a leftover from 3.5 when we had a concurrent merger, now we have a serial one?

ayende replied:
very likely, this also had the issue of the deadlock

grisha-kotler committed:
RavenDB-15776 - PR comments

ayende commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/TimeTrackingSerialMergeScheduler.cs:
```diff
@@ -0,0 +1,53 @@
+                    while (true)
+                    {
+                        MergePolicy.OneMerge merge = writer.GetNextMerge();
+                        if (merge == null)
```
TBD - put a limit on how much time we allow this to run.

ayende committed:
RavenDB-15776
- Add configuration to limit merge duration and maximum segment size to Merge
- Avoid the option of merging excessively sized segments, which can force us to swap

ppekrol commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -65,6 +65,14 @@ public virtual bool RunInMemory
+        [DefaultValue(60)]
+        [TimeUnit(TimeUnit.Seconds)]
+        [IndexUpdateType(IndexUpdateType.Refresh)]
+        [ConfigurationEntry("Indexing.MaxTimeForMergesToKeepRunning", ConfigurationEntryScope.ServerWideOrPerDatabase)]
```
Indexing.MaxTimeForMergesToKeepRunningInSec

ppekrol replied:
I will fix it

ppekrol committed:
RavenDB-15776 fixing configuration

ppekrol commented:
We have test failures: https://jenkins.hibernatingrhinos.com/job/PR_Tests_v5.0/1640/

grisha-kotler committed:
RavenDB-15776 - add MergeFactor configuration

grisha-kotler committed:
RavenDB-15776 - change default values

grisha-kotler committed:
RavenDB-15776 - add more performance details

grisha-kotler committed:
RavenDB-15776 - ignore SetLength for TempCryptoStream

ppekrol commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -157,6 +160,26 @@ public virtual bool RunInMemory
+        [Description("Expert: The maximum number of segments to process in a single merge")]
+        [DefaultValue(2)]
+        [IndexUpdateType(IndexUpdateType.Refresh)]
+        [ConfigurationEntry("Indexing.MaximumSegmentMergeSizeInMb", ConfigurationEntryScope.ServerWideOrPerDatabase)]
```
This should me: `Indexing.MaximumSegmentsMergeFactor` ?

grisha-kotler replied:
changed

arekpalinski commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -21,6 +22,8 @@ public class IndexingConfiguration : ConfigurationCategory
         {
             _root = root;
+
+            MaximumSegmentMergeSize = new Size(PlatformDetails.Is32Bits ? 128 : 1024, SizeUnit.Megabytes);
```
I didn't expect 1GB to be our default (I was thinking about 4GB). Does it mean that even if we limit the merge size significantly it doesn't affect the querying performance much?

grisha-kotler replied:
This is the minimum size that Lucene is considering for segment merging.
I had 10 segments of 3.9GB each which were merged to a single one of ~40GB.
Resulting in tx of 40GB (40GB of scratch files and ~80GB of compression buffer).
This single transaction took about 3 hours to complete.

arekpalinski replied:
Is `MaximumSegmentMergeSize ` the correct name for that since it's about _min_ (not max)?

grisha-kotler replied:
Sorry, this is the _maximum_ size

arekpalinski commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -157,6 +160,26 @@ public virtual bool RunInMemory
+        [ConfigurationEntry("Indexing.MaximumSegmentMergeSizeInMb", ConfigurationEntryScope.ServerWideOrPerDatabase)]
+        public Size? MaximumSegmentMergeSize { get; protected set; }
+
+        [Description("Expert: The maximum number of segments to process in a single merge")]
```
So 'factor' is number of segments? Is that description correct?

grisha-kotler replied:
It is the max size of a single segment that we consider for merge.

arekpalinski replied:
What's the unit of that size?

arekpalinski commented on src/Raven.Server/Indexing/IndexOutputFilesSummary.cs:
```diff
@@ -1,4 +1,6 @@
 ﻿using System.Runtime.CompilerServices;
+using Raven.Server.Documents.Indexes;
+using Voron.Debugging;
```
Can be removed

grisha-kotler replied:
done

grisha-kotler committed:
RavenDB-15776 - update configuration options

ppekrol commented:
@grisha-kotler can you rebase and fix conflicts?

grisha-kotler committed:
RavenDB-15776 - update Lucene
- PR comments
